File: mobile/lib/features/matches/domain/models/match_model.dart
import 'package:equatable/equatable.dart';

class MatchModel extends Equatable {
  final String? id;
  final String winnerId;
  final String loserId;
  final String score;
  final DateTime date;
  final bool isTournament;
  final bool isOffPeak;
  final bool isMatchmakingChallenge;

  const MatchModel({
    this.id,
    required this.winnerId,
    required this.loserId,
    required this.score,
    required this.date,
    this.isTournament = false,
    this.isOffPeak = false,
    this.isMatchmakingChallenge = false,
  });

  factory MatchModel.fromJson(Map<String, dynamic> json) {
    return MatchModel(
      id: json['id'] as String?,
      winnerId: json['winnerId'] as String,
      loserId: json['loserId'] as String,
      score: json['score'] as String,
      date: DateTime.parse(json['date'] as String),
      isTournament: json['isTournament'] as bool? ?? false,
      isOffPeak: json['isOffPeak'] as bool? ?? false,
      isMatchmakingChallenge: json['isMatchmakingChallenge'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'winnerId': winnerId,
      'loserId': loserId,
      'score': score,
      'isTournament': isTournament,
      'isOffPeak': isOffPeak,
      'isMatchmakingChallenge': isMatchmakingChallenge,
    };
  }

  @override
  List<Object?> get props => [
    id,
    winnerId,
    loserId,
    score,
    date,
    isTournament,
    isOffPeak,
    isMatchmakingChallenge,
  ];
}

File: backend/src/infrastructure/database/models/MatchModel.ts
import { Schema, model, Document, Types } from 'mongoose';

export interface MatchDocument extends Document {
    tenantId: Types.ObjectId;
    winnerId: Types.ObjectId;
    loserId: Types.ObjectId;
    score: string;
    date: Date;
    isTournament: boolean;
    isOffPeak: boolean;
    isMatchmakingChallenge: boolean;
    metadata?: Record<string, any>;
    createdAt: Date;
    updatedAt: Date;
}

const MatchSchema = new Schema<MatchDocument>(
    {
        tenantId: { type: Schema.Types.ObjectId, ref: 'Tenant', required: true, index: true },
        winnerId: { type: Schema.Types.ObjectId, ref: 'AuthUser', required: true, index: true },
        loserId: { type: Schema.Types.ObjectId, ref: 'AuthUser', required: true, index: true },
        score: { type: String, required: true },
        date: { type: Date, required: true, default: Date.now },
        isTournament: { type: Boolean, default: false },
        isOffPeak: { type: Boolean, default: false },
        isMatchmakingChallenge: { type: Boolean, default: false },
        metadata: { type: Schema.Types.Mixed },
    },
    {
        timestamps: true,
    }
);

// Índice compuesto para búsquedas de historial de enfrentamientos
MatchSchema.index({ tenantId: 1, date: -1 });
MatchSchema.index({ winnerId: 1, loserId: 1 });

export const MatchModel = model<MatchDocument>('Match', MatchSchema);
